### DISCUSSION TITLE
How i can get the last entity of a list?

### DISCUSSION THREAD

**Discussion by Julian1993ARG** [serenity-is/Serenity#7161] [Help]
I need to retrieve the most recent state of an entity. Currently, I am using a masterDetail configuration at the row level, which gives me a list of states; on the client side, I filter this list by date to get the latest state. I want to move this filtering logic to the row configuration. How can I achieve this?

**Comment by LSNicholls** [comment]
I do exactly this -- or at least I think this is what you are saying:  I have a list of status values for each entity in a StatusHistory table.  But I also always want to know the *current* status of the entity.  

So, if this is what you are doing, I handle it with a view.  Once you have the view available, you just join it to the main entity in your row definition, and you have what you need.

Here is how I do it; it may be a little more complicated than you need but that is because I have multiple entities with Status; ItemType is just how they are all kept in the same StatusHistoryTable.  

I am able to do it this way, because I have  "Started" and "Ended" columns, system-populated, with the "Started" column able to be overridden by the user. The row with "Ended" null is always the current status.  I'll show you an alternate method that I use in cases where this is not available:

```
create or alter view [v_CurrentProjectStatus] as
(
select sh.ItemId as ProjectId, 
       sh.StatusId,
       sh.Started,
       sh.StatusDate,
       ps.Code as StatusCode, 
       isnull(sh.StatusDate, sh.Started) as EffectiveStarted
       from StatusHistory sh
join v_ItemTypeList v
on sh.ItemTypeId = v.Id 
and v.Code = 'Project'
join v_ProjectStatusList ps
on sh.StatusId = ps.Id
where Ended is null  
) 
go
```
If I did not have "Started" and "Ended" in the table, or if "Ended" was not null for the latest record, which might be the case in your system, I would do it like this -- this is MS SQL Server, but I imagine that other databases also have windowing function syntax, since this is in the sql ansi standard I believe:

```
create or alter view [v_CurrentProjectStatus]  as
(
select * from (
  select sh.ItemId as ProjectId, 
       sh.StatusId,
       ps.Code as StatusCode, 
       sh.YourDateValue,
       row_number()  
               over (partition by sh.ItemId order by sh.YourDateValue desc) as ProjectStatusIndex
       from StatusHistory sh
  join v_ItemTypeList v
  on sh.ItemTypeId = v.Id 
  and v.Code = 'Project'
  join v_ProjectStatusList ps
  on sh.StatusId = ps.Id
  ) xx
  where ProjectStatusIndex = 1
) 
go
```

**Reply by Julian1993ARG** [reply]
Thanks for your answer, but what I want is to be able to add it to the entity row:

```
Row {
**********
["How i can get this value??, " ForeignKey("States", "StateId"), LeftJoin(jIdState)]
public int lastStateId { get; set; }

[Expression($"{jIdState}.[Description]")]
public string IdStateDescription  { get; set; }
********
}
```
I hope it can be understood, but this way I could get a taste if I need it only the last status or for example in a product that has a price list if I only want the last price uploaded not the list, how can I leave it already configured in the Row.

**Comment by LSNicholls** [comment]
Yes, I understand.  If you look at my example, you'll see StatusCode in the view and also StatusId.  These correspond to the values that you are asking about.  You join the view to your entity in the Row file, and this gives you what you are looking for.

In my case, I actually use a view that includes many other calculated attributes, so I don't join with v_CurrentProjectStatus itself. The attributes view  v_ProjectAttributes holds the values from v_CurrentProjectStatus and others:  

```
select p.Id,
  cast(case when currProjectStat.StatusCode in ('Closed') then 1 else 0 end  as bit) as IsProjectReadonly,
  currProjectStat.StatusCode as ProjectStatusCode, 
  currProposalStat.StatusCode as ProposalStatusCode,
 -- lots more here
from [Project] p
join  v_CurrentProjectStatus currProjectStat
on p.Id = currProjectStat.ProjectId
join v_CurrentProposalStatus currProposalStat
on p.Id = currProposalStat.ProposalId
-- lots more here

```
So, in my ProjectRow.cs, the join with the view looks like this; it's directly on the primary key of the Project row:

```
 [DisplayName("Project Id"), NotNull, IdProperty, Identity, ForeignKey("v_ProjectAttributes", "Id"), LeftJoin("jAttr")]
 public int? Id { get; set; }
```

... and the various calculated attributes in ProjectRow.cs look like this (typically they are read-only):

```
[DisplayName("Readonly Project?"), Expression("jAttr.IsProjectReadonly"),  ReadOnly(true)]        
public bool? IsProjectReadonly { get; set; }

[DisplayName("Project Status"), Expression("jAttr.ProjectStatusCode"), ReadOnly(true), QuickSearch]
public string ProjectStatus { get; set; }

[DisplayName("Proposal Status"), Expression("jAttr.ProposalStatusCode"), ReadOnly(true), QuickSearch ]
public string ProposalStatus { get; set; }
```

Does this make sense to you?

**Comment by javaman63** [comment]
The simplest way to do what you want is to define 2 fields in the Master Table.
 One for the update date (LastDate) and the other for the Last Inserted ID in the Detail.

You can update them via a trigger on the detail table or from the save handler of the detail table.

Example of the Master and DetailTable

Order
-----
OrderId
Number
LastOrderDetailId
LastOrderDetailDate

OrderDetail
OrderDetailId
StateId
Date

Via a trigger on your detail table, you have to update LastOrderDetailId and LastOrderDetailDate on the master table. You can also use the SaveHandler of the detail table to update those field
